// @crit-fumble/worldanvil/dist/client/api.js
/**
 * World Anvil API Client
 *
 * Type-safe client for World Anvil Boromir API endpoints.
 */
import { HttpClient } from './http.js';
/**
 * World Anvil API Client
 *
 * @example
 * ```typescript
 * import { WorldAnvilClient } from '@crit-fumble/worldanvil/client'
 *
 * const client = new WorldAnvilClient({
 *   applicationKey: process.env.WORLDANVIL_APP_KEY!,
 *   authToken: process.env.WORLDANVIL_AUTH_TOKEN!,
 * })
 *
 * // Get current user identity
 * const user = await client.identity()
 *
 * // Get user's worlds
 * const { entities: worlds } = await client.worlds.list()
 *
 * // Get articles in a world
 * const { entities: articles } = await client.articles.listByWorld(worldId)
 *
 * // Create a session note
 * const note = await client.notes.create({
 *   title: 'Session 5 Notes',
 *   notesection: { id: sectionId },
 *   content: 'The party ventured into the dungeon...',
 * })
 * ```
 */
export class WorldAnvilClient {
    http;
    // Namespaced API methods
    users;
    worlds;
    articles;
    notebooks;
    notesections;
    notes;
    categories;
    maps;
    markers;
    timelines;
    histories;
    images;
    secrets;
    rpgSystems;
    constructor(config) {
        this.http = new HttpClient(config);
        // Initialize namespaced APIs
        this.users = new UsersApi(this.http);
        this.worlds = new WorldsApi(this.http);
        this.articles = new ArticlesApi(this.http);
        this.notebooks = new NotebooksApi(this.http);
        this.notesections = new NotesectionsApi(this.http);
        this.notes = new NotesApi(this.http);
        this.categories = new CategoriesApi(this.http);
        this.maps = new MapsApi(this.http);
        this.markers = new MarkersApi(this.http);
        this.timelines = new TimelinesApi(this.http);
        this.histories = new HistoriesApi(this.http);
        this.images = new ImagesApi(this.http);
        this.secrets = new SecretsApi(this.http);
        this.rpgSystems = new RpgSystemsApi(this.http);
    }
    /**
     * Update client configuration
     */
    configure(config) {
        this.http.configure(config);
    }
    /**
     * Get current authenticated user identity
     */
    async identity() {
        const { data } = await this.http.get('/identity');
        return data;
    }
}
// ============================================
// Users API
// ============================================
class UsersApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get user by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/user', {
            params: { id },
            granularity,
        });
        return data;
    }
}
// ============================================
// Worlds API
// ============================================
class WorldsApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get world by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/world', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List worlds for the current user
     */
    async list() {
        const { data } = await this.http.get('/user/worlds');
        return data;
    }
    /**
     * Create a new world
     */
    async create(request) {
        const { data } = await this.http.put('/world', {
            body: request,
        });
        return data;
    }
    /**
     * Update a world
     */
    async update(id, request) {
        const { data } = await this.http.patch('/world', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a world
     */
    async delete(id) {
        const { data } = await this.http.delete('/world', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Articles API
// ============================================
class ArticlesApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get article by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/article', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List articles in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/articles', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new article
     */
    async create(request) {
        const { data } = await this.http.put('/article', {
            body: request,
        });
        return data;
    }
    /**
     * Update an article
     */
    async update(id, request) {
        const { data } = await this.http.patch('/article', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete an article
     */
    async delete(id) {
        const { data } = await this.http.delete('/article', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Notebooks API
// ============================================
class NotebooksApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get notebook by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/notebook', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List notebooks in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/notebooks', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new notebook
     */
    async create(request) {
        const { data } = await this.http.put('/notebook', {
            body: request,
        });
        return data;
    }
    /**
     * Update a notebook
     */
    async update(id, request) {
        const { data } = await this.http.patch('/notebook', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a notebook
     */
    async delete(id) {
        const { data } = await this.http.delete('/notebook', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Note Sections API
// ============================================
class NotesectionsApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get note section by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/notesection', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List note sections in a notebook
     */
    async listByNotebook(notebookId) {
        const { data } = await this.http.get('/notebook/notesections', {
            params: { id: notebookId },
        });
        return data;
    }
    /**
     * Create a new note section
     */
    async create(request) {
        const { data } = await this.http.put('/notesection', {
            body: request,
        });
        return data;
    }
    /**
     * Update a note section
     */
    async update(id, request) {
        const { data } = await this.http.patch('/notesection', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a note section
     */
    async delete(id) {
        const { data } = await this.http.delete('/notesection', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Notes API
// ============================================
class NotesApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get note by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/note', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List notes in a note section
     */
    async listBySection(notesectionId) {
        const { data } = await this.http.get('/notesection/notes', {
            params: { id: notesectionId },
        });
        return data;
    }
    /**
     * Create a new note
     */
    async create(request) {
        const { data } = await this.http.put('/note', {
            body: request,
        });
        return data;
    }
    /**
     * Update a note
     */
    async update(id, request) {
        const { data } = await this.http.patch('/note', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a note
     */
    async delete(id) {
        const { data } = await this.http.delete('/note', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Categories API
// ============================================
class CategoriesApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get category by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/category', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List categories in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/categories', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new category
     */
    async create(request) {
        const { data } = await this.http.put('/category', {
            body: request,
        });
        return data;
    }
    /**
     * Update a category
     */
    async update(id, request) {
        const { data } = await this.http.patch('/category', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a category
     */
    async delete(id) {
        const { data } = await this.http.delete('/category', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Maps API
// ============================================
class MapsApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get map by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/map', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List maps in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/maps', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new map
     */
    async create(request) {
        const { data } = await this.http.put('/map', {
            body: request,
        });
        return data;
    }
    /**
     * Update a map
     */
    async update(id, request) {
        const { data } = await this.http.patch('/map', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a map
     */
    async delete(id) {
        const { data } = await this.http.delete('/map', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Markers API
// ============================================
class MarkersApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get marker by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/marker', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List markers on a map
     */
    async listByMap(mapId) {
        const { data } = await this.http.get('/map/markers', {
            params: { id: mapId },
        });
        return data;
    }
    /**
     * Create a new marker
     */
    async create(request) {
        const { data } = await this.http.put('/marker', {
            body: request,
        });
        return data;
    }
    /**
     * Update a marker
     */
    async update(id, request) {
        const { data } = await this.http.patch('/marker', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a marker
     */
    async delete(id) {
        const { data } = await this.http.delete('/marker', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Timelines API
// ============================================
class TimelinesApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get timeline by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/timeline', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List timelines in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/timelines', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new timeline
     */
    async create(request) {
        const { data } = await this.http.put('/timeline', {
            body: request,
        });
        return data;
    }
    /**
     * Update a timeline
     */
    async update(id, request) {
        const { data } = await this.http.patch('/timeline', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a timeline
     */
    async delete(id) {
        const { data } = await this.http.delete('/timeline', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Histories API
// ============================================
class HistoriesApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get history event by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/history', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List history events in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/histories', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new history event
     */
    async create(request) {
        const { data } = await this.http.put('/history', {
            body: request,
        });
        return data;
    }
    /**
     * Update a history event
     */
    async update(id, request) {
        const { data } = await this.http.patch('/history', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a history event
     */
    async delete(id) {
        const { data } = await this.http.delete('/history', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// Images API
// ============================================
class ImagesApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get image by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/image', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List images in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/images', {
            params: { id: worldId },
        });
        return data;
    }
}
// ============================================
// Secrets API
// ============================================
class SecretsApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get secret by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/secret', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List secrets in a world
     */
    async listByWorld(worldId) {
        const { data } = await this.http.get('/world/secrets', {
            params: { id: worldId },
        });
        return data;
    }
    /**
     * Create a new secret
     */
    async create(request) {
        const { data } = await this.http.put('/secret', {
            body: request,
        });
        return data;
    }
    /**
     * Update a secret
     */
    async update(id, request) {
        const { data } = await this.http.patch('/secret', {
            params: { id },
            body: request,
        });
        return data;
    }
    /**
     * Delete a secret
     */
    async delete(id) {
        const { data } = await this.http.delete('/secret', {
            params: { id },
        });
        return data;
    }
}
// ============================================
// RPG Systems API
// ============================================
class RpgSystemsApi {
    http;
    constructor(http) {
        this.http = http;
    }
    /**
     * Get RPG system by ID
     */
    async get(id, granularity = 0) {
        const { data } = await this.http.get('/rpgsystem', {
            params: { id },
            granularity,
        });
        return data;
    }
    /**
     * List all available RPG systems
     * Note: Uses POST per World Anvil API design
     */
    async list() {
        const { data } = await this.http.post('/rpgsystems');
        return data;
    }
}
// Re-export HTTP utilities
export { HttpClient, HttpError } from './http.js';
