// @crit-fumble/worldanvil/dist/client/http.js
/**
 * World Anvil HTTP Client
 *
 * Low-level HTTP client for World Anvil API communication.
 * Works in both Node.js and browser environments.
 */
/**
 * HTTP error with response details
 */
export class HttpError extends Error {
    status;
    apiError;
    constructor(message, status, apiError) {
        super(message);
        this.status = status;
        this.apiError = apiError;
        this.name = 'HttpError';
    }
}
const DEFAULT_BASE_URL = 'https://www.worldanvil.com/api/external/boromir';
/**
 * HTTP client for World Anvil API requests
 */
export class HttpClient {
    config;
    constructor(config) {
        this.config = {
            applicationKey: config.applicationKey,
            authToken: config.authToken,
            baseUrl: (config.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, ''),
            timeout: config.timeout ?? 30000,
            headers: config.headers ?? {},
            userAgent: config.userAgent,
        };
    }
    /**
     * Update configuration
     */
    configure(config) {
        if (config.applicationKey !== undefined) {
            this.config.applicationKey = config.applicationKey;
        }
        if (config.authToken !== undefined) {
            this.config.authToken = config.authToken;
        }
        if (config.baseUrl !== undefined) {
            this.config.baseUrl = config.baseUrl.replace(/\/$/, '');
        }
        if (config.timeout !== undefined) {
            this.config.timeout = config.timeout;
        }
        if (config.headers !== undefined) {
            this.config.headers = config.headers;
        }
        if (config.userAgent !== undefined) {
            this.config.userAgent = config.userAgent;
        }
    }
    /**
     * Build URL with query parameters
     */
    buildUrl(path, params) {
        const url = new URL(path, this.config.baseUrl);
        if (params) {
            for (const [key, value] of Object.entries(params)) {
                if (value !== undefined) {
                    url.searchParams.set(key, String(value));
                }
            }
        }
        return url.toString();
    }
    /**
     * Build headers for request
     */
    buildHeaders(options) {
        const headers = {
            'Content-Type': 'application/json',
            'x-application-key': this.config.applicationKey,
            'x-auth-token': this.config.authToken,
            ...this.config.headers,
            ...options?.headers,
        };
        if (this.config.userAgent) {
            headers['User-Agent'] = this.config.userAgent;
        }
        return headers;
    }
    /**
     * Make HTTP request
     */
    async request(method, path, options) {
        // Add granularity to params if specified
        const params = { ...options?.params };
        if (options?.granularity !== undefined) {
            params.granularity = String(options.granularity);
        }
        const url = this.buildUrl(path, params);
        const headers = this.buildHeaders(options);
        const timeout = options?.timeout ?? this.config.timeout;
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), timeout);
        try {
            const response = await fetch(url, {
                method,
                headers,
                body: options?.body ? JSON.stringify(options.body) : undefined,
                signal: controller.signal,
            });
            const text = await response.text();
            const data = text ? JSON.parse(text) : {};
            if (!response.ok) {
                const apiError = {
                    error: data.error ?? 'Request failed',
                    message: data.message ?? text,
                    code: response.status,
                };
                throw new HttpError(apiError.message ?? apiError.error, response.status, apiError);
            }
            return {
                data: data,
                status: response.status,
                headers: response.headers,
            };
        }
        catch (error) {
            if (error instanceof HttpError) {
                throw error;
            }
            if (error instanceof Error) {
                if (error.name === 'AbortError') {
                    throw new HttpError('Request timeout', 408);
                }
                throw new HttpError(error.message, 0);
            }
            throw new HttpError('Unknown error', 0);
        }
        finally {
            clearTimeout(timeoutId);
        }
    }
    /**
     * GET request
     */
    async get(path, options) {
        return this.request('GET', path, options);
    }
    /**
     * POST request
     */
    async post(path, options) {
        return this.request('POST', path, options);
    }
    /**
     * PUT request (World Anvil uses PUT for create)
     */
    async put(path, options) {
        return this.request('PUT', path, options);
    }
    /**
     * PATCH request
     */
    async patch(path, options) {
        return this.request('PATCH', path, options);
    }
    /**
     * DELETE request
     */
    async delete(path, options) {
        return this.request('DELETE', path, options);
    }
}
